NPI-3688 Improved SP3 content writer tests, test data, header formatting correctness, and comment handling / passthrough options

treefern opened the pull request:
Also:
- Functions for formatting epoch row headers and SP3 header have been refactored, improving clarity and avoiding the need for trimming output before use.

ronaldmaj reviewed (changes requested):
This is looking good and introduces some very handy functionality. Just a couple suggestions for type hints and unittests for new functions. If that is resolved, I think this can go in (2025-02-05: may also need to resolve conflicts in `gnssanalysis/gn_utils.py`)

treefern commented on gnssanalysis/gn_io/sp3.py:
```diff
@@ -226,7 +237,126 @@ def remove_svs_from_header(sp3_df: _pd.DataFrame, sats_to_remove: set[str]) -> N
+
+def update_sp3_comments(
+    sp3_df: _pd.DataFrame,
+    comment_lines: Union[list[str], None] = None,
```
Apparently, this syntactical distinction is not required in Python >=3.9, thanks to PEP 585: https://peps.python.org/pep-0585/

The type can be written as `list` rather than `List`.

We support nothing older than Python 3.9. Python 3.9 itself will be end-of-life in October this year.

treefern replied:
It's somewhat unclear how many things in the existing codebase should be cleaned up moving forward, given this.

treefern replied:
New ticket created to update the rest of the codebase for this: NPI-3875

treefern commented on gnssanalysis/gn_datetime.py:
```diff
@@ -270,11 +275,71 @@ def mjd2j2000(mjd: _np.ndarray, seconds_frac: _np.ndarray, pea_partials=False) -
     return datetime2j2000(datetime)
 
 
+def j2000_to_igs_dt(j2000_secs: _np.ndarray) -> _np.ndarray:
```
Simple unit tests added

treefern replied:
`j2000_to_igs_dt()` transiently tested by unit tests for both of the above functions.

treefern commented:
Added unit tests for SP3 comment handling functionality.

Fixed a bug in logic meant to automatically add SP3 comment lead-in if missing. It turned out I'd messed up the order of operations when combining concatenation with an inline conditional.

Improved clarity of output from SP3 version check function

Fixed docstrings and a unit test based on one, as the date example didn't match the SP3d standard.

treefern commented:
This should be ready to go in now

ronaldmaj commented on gnssanalysis/gn_io/sp3.py:
```diff
@@ -512,16 +646,55 @@ def check_epoch_counts_for_discrepancies(
     # All good, validation passed.
 
 
+def check_sp3_version(sp3_bytes: bytes, strict: bool = False) -> bool:
```
Possible unit test for this? Perhaps just checking for each version type, and then for incorrect version values

treefern replied:
Unit test added

ronaldmaj commented on gnssanalysis/gn_io/sp3.py:
```diff
@@ -1478,6 +1722,25 @@ def trim_to_first_n_epochs(
     return trim_df(sp3_df, keep_first_delta_amount=time_offset_from_start)
 
 
+def sp3_hlm_trans(
```
There is already a function with this name in the file. One of them will need to be deleted

treefern replied:
Not sure how that happened! I've removed what appears to be the more dated one (no epoch-wise support)

ronaldmaj reviewed (changes requested):
Looking great! I think this is almost done. Just a few comments based on my tests of the new functionality you've introduced

treefern commented on gnssanalysis/gn_io/sp3.py:
```diff
@@ -228,7 +239,130 @@ def remove_svs_from_header(sp3_df: _pd.DataFrame, sats_to_remove: set[str]) -> N
+    return output_lines
+
+
+def get_sp3_comments(sp3_df: _pd.DataFrame) -> list[str]:
```
Simple standalone test added (it still leverages `read_sp3()` though)

treefern commented on gnssanalysis/gn_io/sp3.py:
```diff
@@ -228,7 +239,130 @@ def remove_svs_from_header(sp3_df: _pd.DataFrame, sats_to_remove: set[str]) -> N
+    return sp3_df.attrs["COMMENTS"]
+
+
+def update_sp3_comments(
```
Almost standalone test written for this

treefern commented on gnssanalysis/gn_io/sp3.py:
```diff
@@ -958,27 +1143,63 @@ def get_unique_epochs(sp3_df: _pd.DataFrame) -> _pd.Index:
 
     :param _pd.DataFrame sp3_df: The DataFrame containing the SP3 data.
+    :param bool output_comment: Write the SP3 comment lines stored with the DataFrame, into the output. Off by default.
+    :param bool strict_mode: More strictly enforce SP3 specification rules (e.g. comments must have a leading space)
```
I'm concerned it might break some workflows unnecessarily. Logging a warning would be ok, but I'm not sure about raising exceptions for things that in some cases are basically cosmetic (comment formatting).

I've put in a first draft of an `off` / `warn` / `raise` enum for strictness and defaulted it to warn.

treefern commented on gnssanalysis/gn_io/sp3.py:
```diff
@@ -228,7 +239,130 @@ def remove_svs_from_header(sp3_df: _pd.DataFrame, sats_to_remove: set[str]) -> N
+            #  - the full line (if the first two chars don't look like a comment)
+            new_lines[i] = SP3_COMMENT_START + (new_lines[i][2:] if new_lines[i][0:2] == "/*" else new_lines[i])
+        if len(new_lines[i]) > SP3_COMMENT_MAX_LENGTH:
+            raise ValueError(f"Comment line too long! Exceeded 80 chars (including lead-in): '{new_lines[i]}'")
```
Oh, specifically if you put the comments in `df.attrs["COMMENTS"]` then read them with `get_sp3_comments()`?
I hadn't intended to validate in that case. The focus was on ensuring the files output were correct.

There may be some value in flagging input files which have non-compliant comments, during reading.
Though if the comments are included when writing back out, the validation will trigger at that point.

ronaldmaj replied:
Yeah, this seems to work fine now after going back and testing it. 
I'm pretty sure last time I used the `update_sp3_comments()` function in order to put in that list of comments above, and it allowed it when each line began with `/*`. 
Anyway, works now, so happy to say this one is resolved

treefern replied:
I resolved an off-by-one error in one of the checks yesterday. I suspect that may have been the original cause of this issue.

ronaldmaj reviewed (dismissed):
This is a significant amount of work and I think is good to go in now. It introduces a bunch of great functionality into the dealing with sp3's, especially around comments.

Thanks heaps for putting the time in to the unit-tests as well. Those can take a while to put together but it'll definitely help us to not fall into any traps. 

I've tested individual functions myself locally, including `get_sp3_comments`, `update_sp3_comments`, `check_sp3_version`, `gen_sp3_header`
I haven't checked every line of the unit tests, but I have read through them and all tests make sense - the unit-tests either test a potential source of error or just make sure the workflow of working with sp3s is fine.

treefern commented:
Noting some of your earlier comments I think I could add checks to `read_sp3()`, to validate SP3 comments then depending on `scrict_mode`, log warnings or raise exceptions about them.

ronaldmaj commented:
> Noting some of your earlier comments I think I could add checks to `read_sp3()`, to validate SP3 comments then depending on `scrict_mode`, log warnings or raise exceptions about them.

Yeah, if you wanted to include those here you can, otherwise another PR following this could include that

treefern commented:
New unified SP3 comment validation added, along with a bunch more unit tests. Some of the new tests duplicate the existing a bit, but for now it shouldn't pose an issue.

ronaldmaj reviewed (approved):
I've tested the new function `validate_sp3_comment_lines()` across the various options and it seems to be working as intended. I like that the function is able to fix minor issues and we get to choose what the validation function returns if issues are fixed - either failing due to the fact the original was in error or passing as the fixed comments pass the test.

Anyway, I've had a read through the unit test for this as well and it seems to cover all cases.
I think this is good to go in now 👍
